### PR TITLE
More flexibility to repeat failed tasks

### PR DESCRIPTION
**Pull request by o-alexandre-felipe**
## Motivation

I have some long tasks running in a machine that will be turned off tomorrow, I would like to be able to restart any scripts I am running without wasting time executing jobs already done.
With this version of run.pl, I can have this by setting `$cmd=run.pl --pick incomplete`

Sometimes when a script fails, or I interrupt it, I need to look for log messages in the script source and compare to the log in order to figure out what stage I should pass to the script. With `--pick incomplete` if the heavy lifting work is done using parallelization it will basically skip to the point where it stopped before.

Also explained on kaldi-help post [How to execute a program for failed jobs](https://groups.google.com/d/msgid/kaldi-help/d177519f-cce9-49c6-ac36-1eaca05da733o%40googlegroups.com?utm_medium=email&utm_source=footer) sometimes bugs may affect only a few jobs, and in that case it is desirable to run just the failed ones, as mentioned here

## Tests

After I implemented I validated the script as follows

Test task consists in a script that takes one second to finish and failes with probability 5%

```bash
cat <<EOF > unstable.sh
sleep 1;
echo "JOB \$1"
if [ \$((\$RANDOM % 20)) = "0" ] ; then
  echo "JOB \$1 Failed!"
  exit 1;
else
  echo "JOB \$1 Succeded!"
  exit 0;
fi
EOF
chmod +x unstable.sh
```

Using the standard run.pl regardless of previous executions all the jobs will be processed

```bash
attempts=0;
while ! $P/run.pl --max-jobs-run 100 JOB=1:50 run/log.JOB.log bash ./unstable.sh JOB ; do
  attempts=$(($attempts+1))
done
```

    run.pl: 3 / 50 failed, log is in run/log.*.log
    run.pl: 7 / 50 failed, log is in run/log.*.log
    run.pl: 6 / 50 failed, log is in run/log.*.log
    run.pl: 4 / 50 failed, log is in run/log.*.log
    run.pl: 3 / 50 failed, log is in run/log.*.log
    run.pl: 1 / 50 failed, log is in run/log.*.log

The new implementation, by default, does the same as before

```bash
attempts=0;
while ! $P/run2.pl --max-jobs-run 100 JOB=1:50 run/log.JOB.log bash ./unstable.sh JOB ; do
  attempts=$(($attempts+1))
done
```

    run.pl: 1 / 50 failed, log is in run/log.*.log
    run.pl: 4 / 50 failed, log is in run/log.*.log
    run.pl: 4 / 50 failed, log is in run/log.*.log
    run.pl: 2 / 50 failed, log is in run/log.*.log
    run.pl: 2 / 50 failed, log is in run/log.*.log
    run.pl: 1 / 50 failed, log is in run/log.*.log
    run.pl: 3 / 50 failed, log is in run/log.*.log
    run.pl: 1 / 50 failed, log is in run/log.*.log

We have the option to run only jobs that failed (based on the log files) using `--pick failed ` switch

```bash
rm -rf run-failed
attempts=0;
while ! $P/run2.pl --max-jobs-run 100 --pick failed JOB=1:1000 run-failed/log.JOB.log bash ./unstable.sh JOB ; do
  attempts=$(($attempts+1))
done
```

    run.pl: 48 / 1000 failed, log is in run-failed/log.*.log
    run.pl: 3 / 1000 failed, log is in run-failed/log.*.log

We have the option to run just incomplete jobs using switch `--pick incomplete`. An incomplete job is one whose corresponding file is missing, or an interrupted job, that is missing the line with exit code at the end of the log file.

```bash
rm -rf run-incomplete
time $P/run2.pl --max-jobs-run 1 --pick incomplete JOB=1:10 run-incomplete/log.JOB.log bash ./unstable.sh JOB
time $P/run2.pl --max-jobs-run 1 --pick incomplete JOB=1:20 run-incomplete/log.JOB.log bash ./unstable.sh JOB
```

    
    real	0m11.852s
    user	0m0.183s
    sys	0m0.005s
    
    real	0m12.176s
    user	0m0.186s
    sys	0m0.023s

Then I override the original script and created the relevant commit

```bash
mv $P/run{2,}.pl
```

### CODE REVIEW

**Comment by o-alexandre-felipe** [comment]
Note: I was expecting just one commit.
I think the problem is that I created the branch based on my master. The important commit is https://github.com/kaldi-asr/kaldi/pull/4219/commits/2521db336227d2aac82c5c5b1f47a6c87fb6e8d3

File: egs/wsj/s5/utils/parallel/run.pl [hunk 1]
```diff
@@ -153,6 +158,34 @@
   }
 }
 
+sub pick_or_exit {
+  if($job_pick eq 'all'){
```
Comment (danpovey) [inline egs/wsj/s5/utils/parallel/run.pl:162]: I think I'm OK with the PR, but can you please add documentation for this function?

Comment (o-alexandre-felipe) [reply egs/wsj/s5/utils/parallel/run.pl:188]: Thank you for the feedback, please check the documentation I just added.

**Comment by danpovey** [comment]
Thanks!   I'll merge.

**Comment by danpovey** [comment]
Incidentally, another mechanicm to repeat failed tasks is to use run.pl retry.pl xxx instead of run.pl xxx

**Pull request merged**